crates/proof-of-sql/src/sql/proof_gadgets/membership_check.rs

```diff
@@ -0,0 +1,144 @@
+use crate::{
+    base::{database::Column, proof::ProofError, scalar::Scalar, slice_ops},
+    sql::{
+        proof::{
+            FinalRoundBuilder, FirstRoundBuilder, SumcheckSubpolynomialType, VerificationBuilder,
+        },
+        proof_plans::{fold_columns, fold_vals},
+    },
+};
+use alloc::{boxed::Box, vec, vec::Vec};
+use bumpalo::Bump;
+use num_traits::{One, Zero};
+
+/// Perform first round evaluation of the membership check.
+#[allow(dead_code)]
+pub(crate) fn first_round_evaluate_membership_check<'a, S: Scalar>(
+    builder: &mut FirstRoundBuilder<'a, S>,
+    multiplicities: &[usize],
+    alloc: &'a Bump,
+) {
+    let casted_multiplicities: Vec<i128> = multiplicities.iter().map(|&x| x as i128).collect();
+    let alloc_multiplicities = alloc.alloc_slice_copy(casted_multiplicities.as_slice());
+    builder.produce_intermediate_mle(alloc_multiplicities as &[_]);
+    builder.request_post_result_challenges(2);
+}
+
+/// Perform final round evaluation of the membership check.
+#[allow(dead_code)]
+#[allow(clippy::too_many_arguments)]
+pub(crate) fn final_round_evaluate_membership_check<'a, S: Scalar>(
+    builder: &mut FinalRoundBuilder<'a, S>,
+    alloc: &'a Bump,
+    alpha: S,
+    beta: S,
+    columns: &[Column<'a, S>],
+    candidate_subset: &[Column<'a, S>],
+    multiplicities: &[usize],
+    num_rows: usize,
+    candidate_num_rows: usize,
+) {
+    let casted_multiplicities: Vec<i128> = multiplicities.iter().map(|&x| x as i128).collect();
+    let alloc_multiplicities = alloc.alloc_slice_copy(casted_multiplicities.as_slice());
+
+    // Fold the columns
+    let input_ones = alloc.alloc_slice_fill_copy(num_rows, true);
+    let candidate_ones = alloc.alloc_slice_fill_copy(candidate_num_rows, true);
```

Comment:
Making these parameters allows them to be reused, which is likely, saving memory and sumcheck perf.